allennlp/modules/contextual_encoder/contextual_encoder.py

```diff
@@ -0,0 +1,171 @@
+from typing import Optional, Callable, Dict
+
+import torch
+
+from allennlp.common import Registrable
+from allennlp.modules.contextual_encoder.character_encoder import CharacterEncoder
+from allennlp.modules.masked_layer_norm import MaskedLayerNorm
+from allennlp.modules.seq2seq_encoders import Seq2SeqEncoder
+from allennlp.modules.token_embedders import TokenEmbedder
+
+class ContextualEncoder(torch.nn.Module, Registrable):
```

Comment:
How does a `ContextualEncoder` get used?  Where does it get its inputs?  From a `TextField`?  I'm just worried about there being too many options for going from text to embeddings, and that the ways they are used are mutually exclusive, so you have to change how you process your data if you want to switch between them, which breaks one of the main points of the library.

Similarly, the `CharacterEncoder` looks like a fancier version of our [`CnnEncoder`](https://github.com/allenai/allennlp/blob/master/allennlp/modules/seq2vec_encoders/cnn_encoder.py) wrapped inside a `TokenCharactersEmbedder`.  It's great to have a better version of this, except now we have two ways of going through the code to get almost the same result.

It seems like we're getting a proliferation of abstractions.  I _think_ the goal of this one is to have something that wraps a `TextFieldEmbedder` and then a `Seq2SeqEncoder`, adding a few options along the way.  That's great, I see no issue there.  But the way it currently is, it doesn't actually use the `TextFieldEmbedder`, it just duplicates _almost_ the same functionality in an incompatible way, so we can't actually use things interchangeably anymore.

Reply:
I mean, another option here is to just add a `ContextualizedTextFieldEmbedder` as a subclass of `TextFieldEmbedder`, where you group a set of `TokenEmbedders` with a `Seq2SeqEncoder`.

Reply:
Sorry to put all of these comments here, too, but the `TokenCharactersEmbedder` already handles the mask, and the `CharacterEncoder` here doesn't actually use the mask, it just returns it.  So if you just soup up our current `CnnEncoder` with highway layers, you could remove the whole masking logic and the dictionary return type.

Reply:
no, this is good. what you are pointing out is that my approach to the Calypso -> AllenNLP project is too "literal" and that I need to step back and be a bit more holistic about it.

that is, I have been mostly approaching it as "how do I get this code working inside AllenNLP" but I should really be approaching it as "what's the right way to get this code incorporated *as part of* AllenNLP". let me think about that some.

Reply:
Minor addendum to what I said above about the `CnnEncoder` - I wouldn't modify our existing one, as that would break our saved models.  But you could add a `CnnHighwayEncoder`, or something, that's basically your `CharacterEncoder` without the mask, and use that inside a `TokenCharactersEmbedder` to get the same effect.

Reply:
I am really struggling with the right way to do this

* TokenCharactersEncoder doesn't quite work, because it does character_ids -> embedding [add a dim] -> seq2vec_encoder [drop a dim] -> outputs

whereas ContextualEncoder wants to do character_ids -> character_encoder (which is really a seq2seq encoder) -> contextual seq2seq encoder -> outputs

--

so one possibility is to make the CharacterEncoder/CnnHighwayEncoder a TokenEmbedder, then the ContextualEncoder itself a TokenEmbedder that wraps another TokenEmbedder? (since that's really what it is)

the usage inside our current ELMo model will still be nonstandard, but then things fit more nicely into the framework?